engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java

```diff
@@ -0,0 +1,800 @@
+
+      if (!groupMap.containsKey(qualifiedName)) {
+        groupMap.put(qualifiedName, new Group(new ArrayList<>(), comparator, compareFunc));
+      }
+      var group = groupMap.get(qualifiedName);
+      group.elems.add(elem);
+    }
+
+    // Sort groups by the FQN of their comparator, with the default comparator
+    // being the first one (the first group).
+    String defCompFQN = getDefaultComparatorQualifiedName();
+    return groupMap.entrySet().stream()
+        .sorted(
+            (entry1, entry2) -> {
+              var fqn1 = entry1.getKey();
+              var fqn2 = entry2.getKey();
+              if (fqn1.equals(defCompFQN)) {
+                return -1;
+              } else if (fqn2.equals(defCompFQN)) {
+                return 1;
+              } else {
+                return fqn1.compareTo(fqn2);
+              }
+            })
+        .map(Entry::getValue)
+        .collect(Collectors.toList());
+  }
+
+  private Object attachWarnings(Object vector, Set<String> warnings) {
+    var warnArray =
+        warnings.stream()
+            .map(Text::create)
+            .map(text -> Warning.create(EnsoContext.get(this), text, this))
+            .toArray(Warning[]::new);
+    return WithWarnings.appendTo(vector, new ArrayRope<>(warnArray));
+  }
+
+  private Object attachDifferentComparatorsWarning(Object vector, List<Group> groups) {
+    var diffCompsMsg =
+        groups.stream()
+            .map(Group::comparator)
+            .map(comparator -> comparator.getQualifiedName().toString())
+            .collect(Collectors.joining(", "));
+    var text = Text.create("Different comparators: [" + diffCompsMsg + "]");
+    var warn = Warning.create(EnsoContext.get(this), text, this);
+    return WithWarnings.appendTo(vector, new ArrayRope<>(warn));
+  }
+
+  private String getDefaultComparatorQualifiedName() {
+    return EnsoContext.get(this)
+        .getBuiltins()
+        .defaultComparator()
+        .getType()
+        .getQualifiedName()
+        .toString();
+  }
+
+  /** A group is "primitive" iff its comparator is the default comparator. */
+  private boolean isPrimitiveGroup(Group group) {
+    return group
+        .comparator
+        .getQualifiedName()
+        .toString()
+        .equals(getDefaultComparatorQualifiedName());
+  }
+
+  private Object incomparableValuesError(Object left, Object right) {
+    return EnsoContext.get(this).getBuiltins().error().makeIncomparableValues(left, right);
+  }
+
+  /**
+   * Helper slow-path method to conveniently gather elements from interop arrays into a java list
+   */
+  @SuppressWarnings("unchecked")
+  private <T> List<T> readInteropArray(
+      InteropLibrary interop, WarningsLibrary warningsLib, Object vector) {
+    try {
+      int size = (int) interop.getArraySize(vector);
+      List<T> res = new ArrayList<>(size);
+      for (int i = 0; i < size; i++) {
+        Object elem = interop.readArrayElement(vector, i);
```

Comment:
Possibly missing conversion to Enso values (https://github.com/enso-org/enso/pull/6280#discussion_r1167372095) - maybe we should have a generic `AtNode` to read an Enso value from anything array-like.

Reply:
Done. See https://github.com/enso-org/enso/pull/5998#discussion_r1167634829